src/v/cloud_roles/refresh_credentials.cc

```diff
@@ -0,0 +1,247 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "cloud_roles/refresh_credentials.h"
+
+#include "cloud_roles/logger.h"
+#include "config/node_config.h"
+#include "model/metadata.h"
+#include "s3/client.h"
+
+#include <seastar/core/abort_source.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/loop.hh>
+
+/// These environment variables can be used to override the default hostname and
+/// port for fetching temporary credentials for testing.
+struct override_api_endpoint_env_vars {
+    static constexpr std::string_view host = "RP_SI_CREDS_API_HOST";
+    static constexpr std::string_view port = "RP_SI_CREDS_API_PORT";
+};
+
+/// Multiplier to derive sleep duration from expiry time. Leaves 0.1 * expiry
+/// seconds as buffer to make API calls. For default expiry of 1 hour, this
+/// results in a fetch after 54 minutes.
+static constexpr float sleep_from_expiry_multiplier = 0.9;
+
+cloud_roles::refresh_credentials::refresh_credentials(
+  std::unique_ptr<impl> impl,
+  ss::gate& gate,
+  ss::abort_source& as,
+  credentials_update_cb_t creds_update,
+  s3::aws_region_name region)
+  : _impl(std::move(impl))
+  , _gate(gate)
+  , _as(as)
+  , _credentials_update(std::move(creds_update))
+  , _region{std::move(region)} {}
+
+void cloud_roles::refresh_credentials::start() {
+    ssx::background = ssx::spawn_with_gate_then(
```

Comment:
I'm not understanding where the gate is closed. presumably the gate is controlled by some higher level?

this is generally ok and we have cases like this in the tree, but a design pattern that is easier to understand, when it can be used, is to make every object that starts background work contain a gate/abort_source and be responsible for its own background operations via start/stop interfaces. then that is external contract rather than requiring passing down references to gates/abort sources.

Reply:
this gate is closed here https://github.com/redpanda-data/redpanda/blob/dev/src/v/cloud_storage/remote.cc#L157 - it is borrowed as reference from `remote` along with abort source and our lifecycle is implicitly tied to remote.

This is a little hard to manage, the other design you mentioned sounds good, I will look into creating and managing the gate and abort source within the refresh object and exposing methods to start/stop and have remote call them, although possibly in a future PR if that is ok? 

If not I can take a pass at implementing them here.